### ISSUE TITLE
array-callback-return: Allow any return statement instead of just top-level ones

### ISSUE THREAD

**Issue by silverwind** [eslint/eslint#9781]
**Tell us about your environment**

* **ESLint Version:** 4.14.0
* **Node Version:** 9.3.0
* **npm Version:** 5.6.0

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

```yaml
rules:
  array-callback-return: [2]
```

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
arr.some(el => {
  if (condition) {
    // do something once
    return true;
  }
})
```

**What did you expect to happen?**

The code to pass.

**What actually happened? Please include the actual, raw output from ESLint.**

```
  1:13  error  Expected to return a value at the end of arrow function
```

The rule currenly checks for the presence of a top-level `return` statement at the end of the function body which isn't useful in case one wants to utilize the implicit `undefined` return value of a function.

I like using a conditional `return` on Array prototype methods like `some`, `every`, `find` to do something and early-exit from the loop once a condition is met. To fix above example, I'd have to add a unnecessary `return false` at the end of the function body.

Suggestion would be to check for the presence of **any** `return` statement inside the function body or introduce an option for such an behaviour.

**Comment by platinumazure** [comment]
Hi @silverwind, thanks for the issue.

I think we could probably consider an enhancement request here. Could you please follow the instructions in the eslint bot's post, and then we can go from there? Thanks!

**Comment by not-an-aardvark** [comment]
Does the `allowImplicit` option solve your use case?

**Comment by silverwind** [comment]
> Does the allowImplicit option solve your use case?

No, that option only allows bare `return`, the rule still checks for it at the top-level.

**Comment by silverwind** [comment]
**What rule do you want to change?**

[array-callback-return](https://eslint.org/docs/rules/array-callback-return)

**Does this change cause the rule to produce more or fewer warnings?**

Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**

I think the default behaviour should be changed. The aim of this rule seems to be to point out missing `return` statements, for example a common mistake is `Array.prototype.map` without a `return` statement which can be refactored to `Array.prototype.forEach`. Right now, the rule looks for return statements only at the top-level, which limits the usefulness of the rule when one wants to utilize the implicit `undefined` return value of a function.

**Please provide some example code that this change will affect:**

````js
arr.some(el => {
  if (condition) {
    // do something once
    return true;
  }
})
````

**What does the rule currently do for this code?**

It gives a warning because it cannot find a top-level `return` statement.

**What will the rule do after it's changed?**

The rule would look for any `return` statement in the function's body, including non-top-level ones, like the one inside the `if` in the example.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
